includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,384 @@
  * @param AbstractObject $object An Islandora Object
  * @param String $filePath video file path
  */
-function islandora_video_streaming_panopto_upload($object, $filePath) {
-  if ($force || !isset[$object['KALTURA_STREAM']) {
+
+
+libraries_load('aws'); 
+
+use Aws\S3\S3Client;
+use Aws\Exception\AwsException;
+use Aws\S3\ObjectUploader;
+use Aws\S3\MultipartUploader;
+use Aws\Exception\MultipartUploadException;
+
+/**
+ * Implement Panopto video upload  
+ * @param AbstractObject $object An Islandora Object
+ * @param string $filePath Path of object
+ * @return array An array that describes the success or failure of the upload 
+ */
+function islandora_video_streaming_panopto_upload($object, $force) {
+  if ($force || !isset($object['PANOPTO_INFO'])) {
+    module_load_include('inc', 'islandora_video', "includes/derivatives" );
     $archival_file = islandora_video_copy_obj_datastream($object);
     if ($archival_file === FALSE) {
       return islandora_video_no_obj_datastream($object->id);
     }
     $filePath = drupal_realpath($archival_file['file']->uri);
     $failureMessage = array(
       'success' => false,
-      'messages' => array(
-        'message' => t('Failed to upload object @pid to Panopto, but no exception caught.'),
+      'messages' => array(),
+    );
+    $streaming = array();
+    
+    $oauthendpoint = variable_get('islandora_video_streaming_panopto_host');
+    $uploadendpoint = $oauthendpoint . '/Panopto/PublicAPI/Rest/sessionUpload';
+    $oauthendpoint .= "/Panopto/oauth2/connect/token";
+    $token = islandora_video_streaming_panopto_get_token($oauthendpoint);
+    $session = null;
+    if ($token) {
+      $session = islandora_video_streaming_panopto_get_session($token, $uploadendpoint); 
+    }
+    $success = false;
+    if ($session) {
+      $upload_id = $session['ID'];
+      $upload_target = $session['UploadTarget'];
+      $manifest_file = islandora_video_streaming_panopto_create_manifest_for_video($filePath);
```

Comment:
`$manifest_file` is a temporary file, and should be cleaned up when leaving this function.